### PR TITLE
ARM split - Add uart half duplex transport support

### PR DESCRIPTION
**Pull request by zvecr**
## Current Issue
Transport does not exist

## TL;DR
This PR adds a hardware backed transport mechanism which *might* work on I2C based boards.

## Description
Provides a method of transport that does not require a patch or upgrade to ChibiOS. Hardware backed, this avoids some of the locking conflicts from within #7072. The downside is that it requires pullups on the TX line, and using pins which share I2C/USART, rather than the bitbang serial which can run on whatever. A good proportion of pro micro splits can run I2C, so this is potentially already covered for CTPC.

Scan rate goes over 2k with higher speed options so more than enough to keep up with current AVR boards.

TODO
- [x] non f303 testing
- [x] Unpick makefile changes - #7990
- [x] Remove keyboard changes from testing
- [x] Add docs

### Testing Notes

| Chip        | Status                        | Notes |
|--------------|-----------------------------|-|
| f072         | ?                                | compiles but not enough hardware to test |
| f103         | :heavy_check_mark: | with `AFIO->MAPR \|= AFIO_MAPR_USART1_REMAP` |
| f303         | :heavy_check_mark: | |
| f401/f411 | :heavy_check_mark: | |

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by drashna** [comment]
Looks good, though, the default keymaps for the levinson is compiling too large.

**Comment by zvecr** [comment]
Going to rip those extra files out once I rebase for #7995. This PR *should* have no impact on existing board, unless you basically add those changes back locally, but while im fixing things up I thought i would leave them there.

File: drivers/arm/serial.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    uint8_t *target2initiator_buffer;
+} SSTD_t;
+#define TID_LIMIT(table) (sizeof(table) / sizeof(SSTD_t))
+
+// initiator is transaction start side
+void soft_serial_initiator_init(SSTD_t *sstd_table, int sstd_table_size);
+// target is interrupt accept side
+void soft_serial_target_init(SSTD_t *sstd_table, int sstd_table_size);
+
+// initiator resullt
```
Comment (tzarc) [inline drivers/arm/serial.h]: `resullt` -> `result`

File: drivers/arm/serial_usart.c [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+#    endif
+#endif
+
+#ifndef USART_DRIVER
+#    define USART_DRIVER SD1
+#endif
+
+#ifndef SELECT_SOFT_SERIAL_SPEED
+#    define SELECT_SOFT_SERIAL_SPEED 1
+// TODO: correct speeds...
```
Comment (tzarc) [inline drivers/arm/serial_usart.c]: Lets just drop the comments, they're listed just below.

File: drivers/arm/serial_usart.c [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+static inline void sdClear(SerialDriver* driver) {
+    while (sdGetTimeout(driver, TIME_IMMEDIATE) != MSG_TIMEOUT) {
+        // Do nothing with the data
+    }
+}
+
+static inline msg_t sdPutHalfDuplex(SerialDriver* driver, uint8_t data) {
+    msg_t ret = sdPut(driver, data);
+    sdGet(driver);
```
Comment (tzarc) [inline drivers/arm/serial_usart.c]: Lets add `// Half duplex requires us to read back the data we just wrote - just throw it away` to match the other functions.

**Comment by tzarc** [comment]
I guess half of the requested changes should probably be split out into a differently-focused PR, specifically that works for both the AVR and ARM implementations. Split transport uses the `SSTD_t` stuff, so probably best left out of scope for this PR. Will go cancel the requested changes accordingly.

File: docs/serial_driver.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+### USART Half-duplex
+Targeting STM32 boards where communication is offloaded to a USART hardware device. The advantage is that this provides accurate fas t and accurate timings. `SOFT_SERIAL_PIN` for this driver is the configured USART TX pin. **The TX pin must have appropriate pull-up resistors**. To configure it, add this to your rules.mk:
+
+```make
+SERIAL_DRIVER = usart
+```
+
+Configure the hardware via your config.h:
+```c
+#define SOFT_SERIAL_PIN B6  // USART TX pin
```
[the remark this answers is not in the document]
Comment (zvecr) [reply docs/serial_driver.md]: Its more that this aims to provide the same interface as the bitbang stuff. For boards where SOFT_SERIAL_PIN is D0, this could be the default implementation. In that case it would have to be driven off the existing SOFT_SERIAL_PIN define.

Personally I dont like the config option of `SOFT_SERIAL_PIN` for any of the implementations, but thats a difficult thing to change now its in the codebase.

File: docs/serial_driver.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+All drivers in this category have the following characteristics:
+* Provides data and signaling over a single conductor
+* Limited to single master, single slave
+
+## Supported Driver Types
+
+|                   | AVR                | ARM                |
+|-------------------|--------------------|--------------------|
+| bit bang          | :heavy_check_mark: | Soon™              |
+| USART Half-duplex |                    | :heavy_check_mark: |
```
[the remark this answers is not in the document]
Comment (zvecr) [reply docs/serial_driver.md]: Is it a hard no? My understanding is it could, but I don't personally want to invest the time into it.

File: drivers/arm/serial_usart.c [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+void soft_serial_target_init(SSTD_t* sstd_table, int sstd_table_size) {
+    Transaction_table      = sstd_table;
+    Transaction_table_size = (uint8_t)sstd_table_size;
+
+    usart_slave_init();
+}
+
+void handle_soft_serial_slave(void) {
+    uint8_t sstd_index = sdGet(&USART_DRIVER);  // first chunk is always transaction id
```
[the remark this answers is not in the document]
Comment (zvecr) [reply drivers/arm/serial_usart.c]: In theory, this code shouldn't be called outside of `transport.c` so *should* be main thread. I can definitely add the locking, but i cant see a use case for it right now? I2C could try and talk to multiple devices, but if that were to happen here it would lead to bad times.

File: drivers/arm/serial_usart.c [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+#ifndef SERIAL_USART_DRIVER
+#    define SERIAL_USART_DRIVER SD1
+#endif
+
+#ifndef SELECT_SOFT_SERIAL_SPEED
+#    define SELECT_SOFT_SERIAL_SPEED 1
+#endif
+
+#ifdef SERIAL_USART_SPEED
```
Comment (tzarc) [inline drivers/arm/serial_usart.c]: Needs `#ifndef`, not `#ifdef`

Comment (tzarc) [reply drivers/arm/serial_usart.c]: Errr, ignore me.

**Comment by tzarc** [comment]
I've tested these at systick frequency of 10k and 100k on an STM32L082, running at 32MHz with a 32MHz USART clock.

As discussed on Discord, this seems to work fine, provided I boost the speed of the UART high enough.
The default baud rate of `115200` seems to produce a link that doesn't want to synchronise both sides. The highest value, `230400` also failed.

The first working value for this device seemed to be `460800`, and the values `921600`, `1843200`, and `2000000` all function fine. Any higher, like `2457600` seems to fail.

For a `handwired/splittest`, I get around a 3650 matrix scan rate at 2M baud.

I'd imagine there's something going on with the pulse width and signalling between devices with respect to the half-duplex mode. It's surprising that it's failing at low baud rates.

**Comment by tzarc** [comment]
PWM-based WS2812 split works fine, too.

**Comment by tzarc** [comment]
Although it initially doesn't sync, hitting RGB_MOD etc. does make things cooperate, though.

**Comment by tzarc** [comment]
As I'd mentioned on Discord, there's a significant number of issues with messages not getting handled correctly on the slave side.

Increasing the baud rate as well as the serial buffer size helped somewhat, but the default set of baud rates end up with pretty consistent failures.

With the pair of L082's the baud rate needed to get up to 2Mbps to get it to be relatively stable, but attempting to send rgblight updates tended to result in failure pretty frequently.

Getting matrix updates did not seem to have any problems at all -- wondering if it's due to minimal data transfer from master to slave.

I think I'd like to try adding some sort of magic prefix and data checksum or something to validate the transfers, retrying if a NACK is received or something.

**Comment by tzarc** [comment]
Actually, first up, I'll try messing about with the stop bits and parity bits and stuff.

File: docs/serial_driver.md [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+### USART Half-duplex
+Targeting STM32 boards where communication is offloaded to a USART hardware device. The advantage is that this provides fast and accurate timings. `SOFT_SERIAL_PIN` for this driver is the configured USART TX pin. **The TX pin must have appropriate pull-up resistors**. To configure it, add this to your rules.mk:
+
+```make
+SERIAL_DRIVER = usart
+```
+
+Configure the hardware via your config.h:
+```c
+#define SOFT_SERIAL_PIN B6  // USART TX pin
```
Comment (drashna) [inline docs/serial_driver.md:47]: Does this default to B6?  If you're using CTPC, this may not be B6.

Comment (zvecr) [reply docs/serial_driver.md:47]: No default, it was just an example based on D0 above, but pins change depending on which stm chip you target so couldnt list them all.

File: drivers/chibios/serial_usart.c [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+    while (true) {
+        handle_soft_serial_slave();
+    }
+}
+
+__attribute__((weak)) void usart_init(void) {
+#if defined(USE_GPIOV1)
+    palSetLineMode(SERIAL_USART_TX_PIN, PAL_MODE_STM32_ALTERNATE_OPENDRAIN);
+#else
+    palSetLineMode(SERIAL_USART_TX_PIN, PAL_MODE_ALTERNATE(SERIAL_USART_TX_PAL_MODE) | PAL_STM32_OTYPE_OPENDRAIN);
```
Comment (tzarc) [inline drivers/chibios/serial_usart.c:121]: CubeMX sets these as push/pull, not open drain.

Comment (zvecr) [reply drivers/chibios/serial_usart.c:121]: Wasnt that for full duplex and not half?

Comment (tzarc) [reply drivers/chibios/serial_usart.c:121]: Just re-confirmed, you're correct. Ignore this.

File: docs/serial_driver.md [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+#define SELECT_SOFT_SERIAL_SPEED 1 // or 0, 2, 3, 4, 5
+                                   //  0: about 230kbps (Experimental only)
+                                   //  1: about 115kbps (default)
+                                   //  2: about 57kbps
+                                   //  3: about 38kbps
+                                   //  4: about 19kbps
+                                   //  5: about 14kbps
```
Comment (tzarc) [inline docs/serial_driver.md:54]:
```suggestion
                                   //  0: about 460800 baud
                                   //  1: about 230400 baud (default)
                                   //  2: about 115200 baud
                                   //  3: about 57600 baud
                                   //  4: about 38400 baud
                                   //  5: about 19200 baud
```

Comment (zvecr) [reply docs/serial_driver.md]: Good catch!
